src/output/plugins/visualization/SoundAnalysis.cxx

```diff
@@ -0,0 +1,394 @@
+  
+               n/N             2*n/N
+    freq_lo * 2  ...freq_lo * 2
+  
+    ...
+               (N-1)*n/N           n
+    freq_lo * 2  ...    freq_lo * 2
+
+ \endcode
+ * 
+ * \subsection vis_out_dft_eg Worked Example
+ * 
+ * Let the number of samples n be 576. This means our dft will return n/2 + 1 =
+ * 289 complex numbers. Let our sampling frequency F be 44,100Hz. For each k,
+ * k=0...289, the corresponding frequency will be k * 44100/289, giving us a
+ * frequency range of 0Hz to 22,500Hz. Let's clamp that to 200-11,000, compute
+ * the power spectrum, and divide that power up into three bands: bass, mids &
+ * trebs.
+ * 
+ * First, we need to find the indicies into the dft corresponding to our
+ * desired frequency range.
+ * 
+ \code
+                       f  * n
+         k * F          k
+    f  = ----- ==> k = ------, where f := the frequency of the k-th
+     k     n              F           k   Fourier coefficient
+  
+  
+         | 200 * 576 |
+    k0 = | --------- | = floor(2.61...) = 2
+         |  44100    |
+          -         -
+  
+          -           -
+         | 11000 * 576 |
+    k1 = | ----------- | = ceil(143.67...) = 144
+         |   44100     |
+  
+ \endcode
+ *
+ * So the power spectrum will have 144 - 2 + 1 = 143 elements in it. Nb. we're
+ * throwing away roughly the upper-half of our frequency spectrum.
+ * 
+ * To divide these frequencies into three bands such that each band contains the
+ * same number of octaves, we compute how many octaves there are in our
+ * frequency range (call this \c n):
+ *
+ \code
+  
+     n   11000        log(11000/200)
+    2  = ----- => n = -------------- = 5.7814
+          200            log(2)
+ \endcode
+ * 
+ * In other words, there are 5.7814 octaves in our chosen frequency range. We
+ * want to give each band 5.7814/3 = 1.9271 octaves. That means the three
+ * "buckets" will be:
+ *
+ \code 
+                           1.9271
+    200 ........... 200 * 2        or   200 -    761Hz
+  
+           1.9271         2*1.9271
+    200 * 2 ....... 200 * 2        or   761 -  2,892Hz
+  
+           2*1.9271        5.7814
+    200 * 2 ....... 200 * 2        or 2,892 - 11,000Hz
+
+ \endcode
+ *
+ *
+ */
+
+Visualization::SoundAnalysis::SoundAnalysis(
+	const SoundAnalysisParameters &params,
+	std::shared_ptr<SoundInfoCache> pc)
+: num_samples(params.GetNumSamples()), 
+  out_samples((num_samples / 2) + 1),
+  pcache(pc),
+  audio_format(pc->GetFormat()),
+  num_channels(audio_format.channels), 
+  cbbuf(params.GetNumSamples() * audio_format.GetFrameSize()), 
+  buf(new std::byte[cbbuf]),
+  in(fftwf_alloc_real(num_samples * num_channels), fftwf_free),
+  out(fftwf_alloc_complex(out_samples * num_channels), fftwf_free),
+  bass_mids_trebs(new float[3 * num_channels])
+{
+	if (num_samples > INT_MAX) {
+		throw FmtInvalidArgument("num_samples({}) may not be larger than {}",
+								 num_samples, INT_MAX);
+	}
+	
+	int n[] = { (int)num_samples };
```

Comment:
Why is this signed?

Reply:
That would be a question for the authors of libfftw :) I'm just conforming to the API.